setup.py

```diff
@@ -21,29 +21,61 @@
 import os
 import platform
 
-from setuptools import setup, Extension, Feature
+
+from distutils.errors import CCompilerError
+from distutils.errors import DistutilsExecError
+from distutils.errors import DistutilsPlatformError
+
+from setuptools import Extension
+from setuptools.command.build_ext import build_ext
+from setuptools import setup
+from setuptools import Feature
+
+class optional_build_ext(build_ext):
+    """This class subclasses build_ext and allows
+       the building of C extensions to fail.
+    """
+    def run(self):
+        try:
+            build_ext.run(self)
+        except DistutilsPlatformError as e:
+            self._unavailable(e)
+
+    def build_extension(self, ext):
+        try:
+            build_ext.build_extension(self, ext)
+        except (CCompilerError, DistutilsExecError, OSError) as e:
+            self._unavailable(e)
+
+    def _unavailable(self, e):
+        print('*' * 80)
+        print("""WARNING:
+        An optional code optimization (C extension) could not be compiled.
+        Optimizations for this package will not be available!""")
+        print()
```

Comment:
On Python 2 this prints an empty tuple.  Use `print("")` or add a `from __future__ import print_function` at the top.

Reply:
Honestly, I've just been cargo-culting this code around from zope.interface (which doesn't have the future import and doesn't print to sys.stderr). So I'm afraid a few projects (zope.security, zope.hookable plus the original zope.interface) will be printing superfluous tuples.

Reply:
It's not a big problem since pip will hide this message from users anyway ;)

zc.buildout users might notice.